packages/services/src/state_listener.rs

```diff
@@ -0,0 +1,294 @@
+use async_trait::async_trait;
+use ports::{
+    storage::Storage,
+    types::{SubmissionTx, TransactionReceipt, TransactionState},
+};
+
+use super::Runner;
+
+pub struct StateListener<L1, Db> {
+    l1_adapter: L1,
+    storage: Db,
+    num_blocks_to_finalize: u64,
+}
+
+impl<L1, Db> StateListener<L1, Db> {
+    pub fn new(l1_adapter: L1, storage: Db, num_blocks_to_finalize: u64) -> Self {
+        Self {
+            l1_adapter,
+            storage,
+            num_blocks_to_finalize,
+        }
+    }
+}
+
+impl<L1, Db> StateListener<L1, Db>
+where
+    L1: ports::l1::Api,
+    Db: Storage,
+{
+    async fn check_pending_txs(&mut self, pending_txs: Vec<SubmissionTx>) -> crate::Result<()> {
+        let current_block_number: u64 = self.l1_adapter.get_block_number().await?.into();
+
+        for tx in pending_txs {
+            let Some(tx_receipt) = self.l1_adapter.get_transaction_receipt(tx.hash).await? else {
+                continue; // not committed
+            };
+
+            // Status: either 1 (success) or 0 (failure).
+            // Only present after activation of [EIP-658](https://eips.ethereum.org/EIPS/eip-658)
+            if let Some(status) = tx_receipt.status {
+                let status: u64 = status.try_into().map_err(|_| {
+                    crate::Error::Other(
+                        "could not convert tx receipt `status` to `u64`".to_string(),
+                    )
+                })?;
+
+                if status == 0 {
+                    self.storage
+                        .update_submission_tx_state(tx.hash, TransactionState::Failed)
+                        .await?;
+                }
+
+                continue;
+            }
```

Comment:
I'm for the option of having our own `TransactionResponse` with only the fields we're interested in.

When we convert it over in the adapter we can do this validation there -- i.e. check that the status is valid and map it into something like a `succeeded: bool`. 

Later on we also validate that the block number is present and that it can be converted into a u64. This can also be moved to the TryFrom step between an ethers transaction receipt and our own.

Would also be prudent to move away from the ethers type since we're already planning on migrating to alloy.

Reply:
added `TransactionResponse` and validation done in adapter